storage/innobase/xtrabackup/src/xtrabackup.cc

```diff
@@ -2156,11 +2162,114 @@ static bool innodb_init_param(void) {
   return (TRUE);
 }
 
+/** Scan the sdi id from DD table "mysql.tables"
+@param[in]  name       name of table database/name
+@param[out] sdi_id     id of table
+@param[in]  thd        THD */
+void get_id_from_dd_scan(const std::string name, uint64 *id, THD *thd) {
+  std::string db_name;
+  std::string table_name;
+  uint64 schema_id = 0;
+
+  /* get the database and table_name from space name */
+  dict_name::get_table(name, db_name, table_name);
+
+  ut_ad(db_name.compare("mysql") != 0);
+
+  dict_table_t *sys_tables = nullptr;
+  btr_pcur_t pcur;
+  const rec_t *rec = nullptr;
+  mtr_t mtr;
+  MDL_ticket *mdl = nullptr;
+  mem_heap_t *heap = mem_heap_create(1000);
+
+  /* map of schema name and id build from scanning mysql/schemata */
+  if (dd_schema_map.size() == 0) {
+    mutex_enter(&(dict_sys->mutex));
+
+    mtr_start(&mtr);
+    rec = dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/schemata",
+                              &sys_tables);
+    while (rec) {
+      uint64 rec_schema_id;
+      std::string rec_name;
+
+      dd_process_schema_rec(heap, rec, sys_tables, &mtr, &rec_name,
+                            &rec_schema_id);
+      dd_schema_map.insert(std::make_pair(rec_name, rec_schema_id));
+      mem_heap_empty(heap);
+
+      mtr_start(&mtr);
+      rec = (rec_t *)dd_getnext_system_rec(&pcur, &mtr);
+    }
+
+    mtr_commit(&mtr);
+    dd_table_close(sys_tables, thd, &mdl, true);
+    mem_heap_empty(heap);
+
+    mutex_exit(&(dict_sys->mutex));
+  }
+
+  /* map of <schema id, name> and key build from scanning mysql/tables */
+  if (dd_table_map.size() == 0) {
+    mutex_enter(&(dict_sys->mutex));
+    mtr_start(&mtr);
+
+    rec = dd_startscan_system(thd, &mdl, &pcur, &mtr, "mysql/tables",
+                              &sys_tables);
+
+    while (rec) {
+      uint64 rec_schema_id;
+      std::string rec_name;
+      uint64 rec_id;
+
+      dd_process_dd_tables_rec(heap, rec, sys_tables, &mtr, &rec_schema_id,
+                               &rec_name, &rec_id);
+      mem_heap_empty(heap);
+
+      auto rec_table_id = std::make_pair(rec_schema_id, rec_name);
+
+      dd_table_map.insert(
+          std::make_pair(std::make_pair(rec_schema_id, rec_name), rec_id));
+
+      mtr_start(&mtr);
+      rec = (rec_t *)dd_getnext_system_rec(&pcur, &mtr);
+    }
+
+    mtr_commit(&mtr);
+    dd_table_close(sys_tables, thd, &mdl, true);
+    mem_heap_free(heap);
+
+    mutex_exit(&(dict_sys->mutex));
+  }
+
+  schema_id = dd_schema_map[db_name];
+
+  ut_ad(schema_id != 0);
+  if (schema_id == 0) {
+    msg("xtrabackup: can't find %s entry in mysql/schemata for tablespace %s\n",
+        db_name.c_str(), name.c_str());
+    exit(EXIT_FAILURE);
```

Comment:
btw, these exits() are a bad idea. I found a feature request that PXB should log the status of a backup. and in case of failure, they want the reason for failure to be logged.

Can you please try to avoid these exits() and do a graceful exit (return until main)

Reply:
fixed. now we see messages like this
xtrabackup: can't find test entry in mysql/schemata for tablespace test/pt1@4f60@597d#p#p1
xtrabackup: innodb_init(): Error occured.
2021-07-29T06:31:17.706469Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'keyring_file'
2021-07-29T06:31:17.706936Z 0 [Note] [MY-010733] [Server] Shutting down plugin 'daemon_keyring_proxy_plugin'

Reply:
:clap: